codex/merkletree/merkletree.nim

```diff
@@ -1,96 +1,136 @@
 ## Nim-Codex
-## Copyright (c) 2022 Status Research & Development GmbH
+## Copyright (c) 2023 Status Research & Development GmbH
 ## Licensed under either of
 ##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
 ##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
 ## at your option.
 ## This file may not be copied, modified, or distributed except according to
 ## those terms.
 
-import std/sequtils
 import std/math
 import std/bitops
-import std/sugar
+import std/strutils
 
-import pkg/libp2p
-import pkg/stew/byteutils
-import pkg/questionable
 import pkg/questionable/results
+import pkg/nimcrypto/sha2
+
+const digestSize = sha256.sizeDigest
 
 type
-  MerkleHash* = MultiHash
+  MerkleHash* = array[digestSize, byte]
   MerkleTree* = object
-    leavesCount: int
+    leavesCount: Natural
     nodes: seq[MerkleHash]
   MerkleProof* = object
-    index: int
+    index: Natural
     path: seq[MerkleHash]
+  MerkleTreeBuilder* = object
+    buffer: seq[MerkleHash]
 
-# Tree constructed from leaves H0..H2 is
-#     H5=H(H3 & H4)
-#    /              \
-#   H3=H(H0 & H1)   H4=H(H2 & H2)
-#  /      \        /
-# H0=H(A) H1=H(B) H2=H(C)
-# |       |       |
-# A       B       C
-#
-# Memory layout is [H0, H1, H2, H3, H4, H5]
-#
-# Proofs of inclusion are
-# - [H1, H4] for A
-# - [H0, H4] for B
-# - [H2, H3] for C
-
+###########################################################
+# Helper functions
+###########################################################
 
 func computeTreeHeight(leavesCount: int): int =
   if isPowerOfTwo(leavesCount): 
     fastLog2(leavesCount) + 1
   else:
     fastLog2(leavesCount) + 2
 
-func getLowHigh(leavesCount, level: int): (int, int) =
-  var width = leavesCount
-  var low = 0
-  for _ in 0..<level:
-    low += width
-    width = (width + 1) div 2
+func computeLevels(leavesCount: int): seq[tuple[offset: int, width: int]] =
+  let height = computeTreeHeight(leavesCount)
+  result = newSeq[tuple[offset: int, width: int]](height)
```

Comment:
We usually avoid using `result` due to a slew of pitfalls. More info here https://status-im.github.io/nim-style-guide/language.result.html

Reply:
Thanks for the info